### PR TITLE
JS cleanup suggestions

### PR DESCRIPTION
**Pull request by khiga8**
- Using custom element conventions such as getters and setters
- Fixing lint rules

### CODE REVIEW

File: app/components/primer/alpha/action-menu-element.ts [hunk 1]
```diff
@@ -8,39 +8,66 @@ export {}
...
-    this.setAttribute('data-anchor-align', value)
-  }
-
   get anchorSide(): AnchorSide {
     return (this.getAttribute('data-anchor-side') || 'outside-bottom') as AnchorSide
   }
 
-  set anchorSide(value: AnchorSide) {
-    this.setAttribute('data-anchor-align', value)
+  get menu(): HTMLUListElement | null {
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:23]: Thanks for doing this! I'm curious what this does as opposed to them being defined at the top?

Comment (khiga8) [reply app/components/primer/alpha/action-menu-element.ts:23]: This makes sure that the variables are only grabbed when we need it, AFTER the element and its children are properly loaded by the browser. I think this relates to why we saw our event handlers weren't working with our documentation site in some cases before we added `<script defer>`

Comment (lindseywild) [reply app/components/primer/alpha/action-menu-element.ts:23]: Ah! Ok wonderful thank you

**Comment by lindseywild** [comment]
This looks good from a cursory glance and some manual testing! It'll be good to have the JS unit tests check this work too 😄

File: app/components/primer/alpha/action-menu-element.ts [hunk 2]
```diff
@@ -8,39 +8,66 @@ export {}
...
+      this.menu.style.visibility = 'hidden'
+
+      this.#updatePosition()
+
+      this.menu.style.visibility = 'visible'
+    } else {
+      if (!this.open) return
+      this.removeAttribute('open')
+      this.trigger?.setAttribute('aria-expanded', 'false')
+      this.#menu && this.#menu.hidden = true
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:56]:
```suggestion
      this.menu && this.menu.setAttribute('hidden', 'true')
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 3]
```diff
@@ -8,39 +8,66 @@ export {}
...
-    this.setAttribute('data-anchor-align', value)
+  get menu(): HTMLUListElement | null {
+    return this.querySelector<HTMLUListElement>('[role="menu"]')
+  }
+
+  get trigger(): HTMLButtonElement | null {
+    return this.querySelector<HTMLButtonElement>('button')
+  }
+
+  get #menuItems(): HTMLElement[] | null {
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:31]:
```suggestion
  get menuItems(): HTMLElement[] | null {
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 4]
```diff
@@ -60,25 +87,32 @@ class ActionMenuElement extends HTMLElement {
...
-    this.#buttonEl.addEventListener('click', this.buttonClick.bind(this), {signal})
+    if (!this.trigger || !this.menu) return
+    this.trigger.addEventListener('keydown', this.buttonKeydown.bind(this), {signal})
+    this.trigger.addEventListener('click', this.buttonClick.bind(this), {signal})
 
-    if (this.#allMenuItemEls) {
-      for (const menuItem of this.#allMenuItemEls) {
-        this.#menuItemEls.push(menuItem)
+    this.#firstCharactersOfItems = []
+    if (this.#menuItems) {
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:114]:
```suggestion
    if (this.menuItems) {
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 5]
```diff
@@ -60,25 +87,32 @@ class ActionMenuElement extends HTMLElement {
...
+      for (const menuItem of this.#menuItems) {
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:115]:
```suggestion
      for (const menuItem of this.menuItems) {
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 6]
```diff
@@ -98,7 +132,9 @@ class ActionMenuElement extends HTMLElement {
   }
 
   setFocusToMenuItem(newMenuItem: HTMLElement) {
-    for (const item of this.#menuItemEls) {
+    if (!this.#menuItems) return
+
+    for (const item of this.#menuItems) {
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:137]:
```suggestion
    if (!this.menuItems) return

    for (const item of this.menuItems) {
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 7]
```diff
@@ -108,23 +144,17 @@ class ActionMenuElement extends HTMLElement {
...
-  setFocusToFirstMenuItem() {
-    this.setFocusToMenuItem(this.#firstMenuItem)
-  }
-
-  setFocusToLastMenuItem() {
-    this.setFocusToMenuItem(this.#lastMenuItem)
-  }
-
   setFocusToPreviousMenuItem(currentMenuItem: HTMLElement) {
+    if (!this.#menuItems) return
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:148]:
```suggestion
    if (!this.menuItems) return
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 8]
```diff
@@ -108,23 +144,17 @@ class ActionMenuElement extends HTMLElement {
...
     let newMenuItem = null
     let index = null
 
     if (currentMenuItem === this.#firstMenuItem) {
       newMenuItem = this.#lastMenuItem
     } else {
-      index = this.#menuItemEls.indexOf(currentMenuItem)
-      newMenuItem = this.#menuItemEls[index - 1]
+      index = this.#menuItems.indexOf(currentMenuItem)
+      newMenuItem = this.#menuItems[index - 1]
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:157]:
```suggestion
      index = this.menuItems.indexOf(currentMenuItem)
      newMenuItem = this.menuItems[index - 1]
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 9]
```diff
@@ -133,21 +163,25 @@ class ActionMenuElement extends HTMLElement {
   }
 
   setFocusToNextMenuItem(currentMenuItem: HTMLElement) {
+    if (!this.#menuItems) return
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:166]:
```suggestion
    if (!this.menuItems) return
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 10]
```diff
@@ -133,21 +163,25 @@ class ActionMenuElement extends HTMLElement {
...
     let newMenuItem = null
     let index = null
 
     if (currentMenuItem === this.#lastMenuItem) {
       newMenuItem = this.#firstMenuItem
     } else {
-      index = this.#menuItemEls.indexOf(currentMenuItem)
-      newMenuItem = this.#menuItemEls[index + 1]
+      index = this.#menuItems.indexOf(currentMenuItem)
+      newMenuItem = this.#menuItems[index + 1]
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:175]:
```suggestion
      index = this.menuItems.indexOf(currentMenuItem)
      newMenuItem = this.menuItems[index + 1]
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 11]
```diff
@@ -133,21 +163,25 @@ class ActionMenuElement extends HTMLElement {
...
     }
     this.setFocusToMenuItem(newMenuItem)
 
     return newMenuItem
   }
 
   setFocusByFirstCharacter(currentMenuItem: HTMLElement, character: string) {
+    if (!this.#menuItems) return
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:183]:
```suggestion
    if (!this.menuItems) return
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 12]
```diff
@@ -158,8 +192,8 @@ class ActionMenuElement extends HTMLElement {
     character = character.toLowerCase()
 
     // Get start index for search based on position of currentMenuItem
-    start = this.#menuItemEls.indexOf(currentMenuItem) + 1
-    if (start >= this.#menuItemEls.length) {
+    start = this.#menuItems.indexOf(currentMenuItem) + 1
+    if (start >= this.#menuItems.length) {
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:196]:
```suggestion
    start = this.menuItems.indexOf(currentMenuItem) + 1
    if (start >= this.menuItems.length) {
```

File: app/components/primer/alpha/action-menu-element.ts [hunk 13]
```diff
@@ -173,53 +207,21 @@ class ActionMenuElement extends HTMLElement {
 
     // If match is found
     if (index > -1) {
-      this.setFocusToMenuItem(this.#menuItemEls[index])
+      this.setFocusToMenuItem(this.#menuItems[index])
```
Comment (lindseywild) [inline app/components/primer/alpha/action-menu-element.ts:210]:
```suggestion
      this.setFocusToMenuItem(this.menuItems[index])
```

**Review by lindseywild — COMMENTED** [review]
Some things, should fix the linting error too. Otherwise this looks good. I tested it locally!

**Pull request merged**